doc/developer/design/20240625_source_versioning__table_from_sources.md

```diff
@@ -0,0 +1,537 @@
+    -   Convert subsources to `CREATE TABLE` statements with the same name
+    -   Change the `CREATE SOURCE` statement to point to the current progress collection and
+        change the source name to `<source>_progress`
+    -   Drop the existing `CREATE SUBSOURCE <source>_progress` statement since this will no
+        longer have a collection to own
+    -   Drop the existing top-level collection tied to the source, since this is already
+        unused and will no longer be owned by any statement
+
+#### Exposing available upstream references to users
+
+Since creating a source will no longer automatically create any subsources, users that
+do not already know all the upstream references available in their system may find it difficult
+to know which `CREATE TABLE .. FROM SOURCE` statements they could use to ingest their
+upstream data. Additionally, the Web Console will need a way to present a dynamic UI
+to users after they have created a source to allow them to create any downstream tables
+from a selection of the available upstream references of that source.
+
+We have two options available to us for exposing this information to users:
+
+1.  Add a new statement to fetch from the upstream all possible 'upstream references' available
+    for ingestion.
+2.  Create a new system table that periodically updates with the set of available upstream
+    references in each sources, and a new statement to trigger a refresh of this system table.
+
+The 2nd option is more complex to implement, but presents the information in a form that
+is more similar to other similar information in Materialize. However since the data can
+become out of date, it's important that the refresh statement is highlighted prominently
+in user-facing documentation and related guides.
+
+The proposal for the 2nd option (a new system table):
+
+Introduce a `mz_catalog.mz_available_source_references` with the following columns:
+
+-   `source_id`: The source (e.g. a Postgres, MySQL, Kafka source) this reference is available for.
+-   `reference`: The 'external reference' for this available object, such that using `reference`
+    in a `CREATE TABLE <table_name> FROM SOURCE <source_id> (REFERENCE <reference>)` would
+    construct a table ingesting this upstream reference. This is essentially the fully-qualified
+    upstream name of the reference.
+-   `columns`: If available, an array of of columns that this external reference would have if it
+    were ingested as a table in Materialize. This will be NULL for source types where an
+    external registry is needed to determine the upstream schema (e.g. Kafka).
+-   `name`: The object name of the reference. E.g. for a Postgres source this is the table name,
+    and for kafka is the topic name.
+-   `namespace`: Any namespace that this reference object exists in, if applicable. For Postgres
+    this would be `<database>.<schema>`, for MySQL this would be `<schema>`, and for Kafka this
+    would be NULL.
```

Comment:
In the past, having a single string containing multiple identifier parts caused all manner of pain for the Terraform provider, around escaping and unescaping and the like. Would strongly encourage having three columns here (`database`, `schema`, `name`, or the like) and no columns for the combined reference. It's slightly less ergonomic from SQL, but it is much, much harder to get escaping wrong in downstream tools.

Reply:
Oh, that's exactly what I was thinking, too!

Reply:
The problem I was trying to avoid is that not all external systems have 2 layers of namespacing, nor do they always call them `database` and `schema` (in mysql those both refer to the single namespace layer, and kafka has no namespace above the topic).

If we include separate columns for each level of possible namespacing, we need to pick names that are not confusing when thinking about the various source-types that this count entail. do you have a suggestion for that?

Reply:
> If we include separate columns for each level of possible namespacing, we need to pick names that are not confusing when thinking about the various source-types that this count entail. do you have a suggestion for that?

Nothing good. 🙈 The only thing that comes to mind besides `database`, `schema`, and `name` is something like `name1`, `name2`, `name3`. Or slightly less bad: `namespace1`, `namespace2`, `name`? Or `high_namespace`, `low_namespace`, `name`?

Reply:
For MySQL, we also made the trade-off in the syntax (`FOR SCHEMAS`, instead of `FOR DATABASES`). PostgreSQL follows the [hierarchy defined in the SQL standard](https://stackoverflow.com/a/17943883), as do SQL Server and Oracle. So...I think it's fine? The table will be in `mz_internal`, so we can always just let it bake for a while and rename the columns at a later point.

Reply:
Taking a detour: we have the same problem with [mz_sources](https://materialize.com/docs/sql/system-catalog/mz_catalog/#mz_sources), currently. It has columns that are common across source types, then three columns that are specific to Kafka sources (envelope_type, key_format, value_format). Should we move these down to mz_kafka_sources?

Reply:
One other alternative design would be to have different references tables per source type:

  * `mz_mysql_source_references`
  * `mz_postgres_source_references`
  * `mz_kafka_source_references`

Then each type of source would be able to use the names for the namespace specific to that type of source.

> Taking a detour: we have the same problem with [mz_sources](https://materialize.com/docs/sql/system-catalog/mz_catalog/#mz_sources), currently. It has columns that are common across source types, then three columns that are specific to Kafka sources (`envelope_type`, `key_format`, `value_format`). Should we move these down to `mz_kafka_sources`?

It's a good question. The `mz_sources` table was designed back in the day when we expected "format" and "envelope" to be common features of ~all source types. Made more sense when we had file and S3 sources that had formats and envelopes. Kafka was the only source type with a key format, of course, but we envisioned adding NATS/Kinesis sources eventually, which would have key formats.

Of course the situation we're in now is that a majority of our source types do not support envelopes or formats of any sort. So it could make sense to start thinking about envelopes and formats as a Kafka-specific thing. 

Changing `mz_sources` like this would be a very backwards incompatible change, though—the kind that has the potential to break our users. We'd probably need to scan the query histories across all accounts to see how people are querying `mz_sources` before pursuing a change like this.

Reply:
> Changing mz_sources like this would be a very backwards incompatible change, though—the kind that has the potential to break our users.

I doubt anyone other than the Console is using these fields, but we can also just deprecate them, rather than remove them!

Reply:
> One other alternative design would be to have different references tables per source type:

This is not a bad idea and would avoid the conflict we have between introducing new columns that are source-type specific, but might be less ergonomic for the console to use for presenting the 'table creation' UI to users. We should discuss this tomorrow